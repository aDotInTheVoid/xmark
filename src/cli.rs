use clap::Clap;
use eyre::Result;

use std::path;

#[derive(Debug)]
pub struct Args {
    pub dir: path::PathBuf,
    pub create: bool,
}

impl Args {
    pub fn parse() -> Result<Self> {
        let ArgsInner { mut dir, create } = ArgsInner::parse();
        dir = dir.canonicalize()?;
        Ok(Args { dir, create })
    }
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: path::PathBuf,
    #[clap(long)]
    pub create: bool,
}
