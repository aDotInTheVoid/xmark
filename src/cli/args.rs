use clap::Clap;
use eyre::Result;
use tracing::instrument;

#[derive(Clap, Debug)]
pub(super) struct Args {
    #[clap(subcommand)]
    subcmd: SubCommand,
}

#[derive(Clap, Debug)]
pub(super) enum SubCommand {
    Build(super::cmd::build::Args), // TODO: init, clean, watch, serve
}

impl Args {
    #[instrument]
    pub fn run(self) -> Result<()> {
        match self.subcmd {
            SubCommand::Build(c) => c.run(),
        }
    }
}
