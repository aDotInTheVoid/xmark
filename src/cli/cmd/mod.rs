//! Individual subcommands.
pub mod build;