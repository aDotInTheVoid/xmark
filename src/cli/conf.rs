//! The xmark.toml file
use serde::{Deserialize, Serialize};
use toml::{self, value::Table};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DiskConf {
    pub books: Vec<Location>,
    #[serde(default)]
    pub output: Table,
}
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum Location {
    Bare(String),
    Named { name: String, root_loc: String },
}

#[cfg(test)]
mod tests {
    use super::*;
    use toml::Value;

    fn check_disc_conf(l: DiskConf, r: &str) {
        assert_eq!(toml::from_str::<DiskConf>(r).unwrap(), l);
    }

    fn l(s: &str) -> Location {
        Location::Bare(s.to_owned())
    }

    #[test]
    fn basic_disc_conf() {
        check_disc_conf(
            DiskConf {
                books: vec![l("hello"), l("w2"), l("x4")],
                output: Table::new(),
            },
            r#"
books = [
    "hello",
    "w2",
    "x4",
]
            "#,
        );
    }

    #[test]
    fn empty_output_html() {
        check_disc_conf(
            DiskConf {
                books: vec![l("just_one")],
                output: vec![("html".to_owned(), Value::Table(Table::new()))]
                    .into_iter()
                    .collect(),
            },
            r#"
books = [ "just_one" ]
[output.html]
            "#,
        )
    }

    #[test]
    fn output_html_vals() {
        let mut html = Table::new();
        html.insert(
            "output-url".to_owned(),
            Value::String("/bookshelf/".to_owned()),
        );
        html.insert("minify".to_owned(), Value::Boolean(false));
        let html = Value::Table(html);

        check_disc_conf(
            DiskConf {
                books: vec![l("the-book")],
                output: vec![("html".to_owned(), html)].into_iter().collect(),
            },
            r#"
books = [ "the-book" ]
[output.html]
output-url = "/bookshelf/"
minify = false
            "#,
        )
    }

    #[test]
    fn multi_output() {
        let mut html = Table::new();
        html.insert(
            "output-url".to_owned(),
            Value::String("/bookshelf/".to_owned()),
        );
        html.insert("minify".to_owned(), Value::Boolean(false));
        let html = Value::Table(html);

        let mut tex = Table::new();
        tex.insert("created".to_owned(), Value::Integer(1978));
        tex.insert("xml".to_owned(), Value::Boolean(false));
        tex.insert("parsing".to_owned(), Value::String("Imposible".to_owned()));
        let tex = Value::Table(tex);

        check_disc_conf(
            DiskConf {
                books: Vec::new(),
                output: vec![("html".to_owned(), html), ("tex".to_owned(), tex)]
                    .into_iter()
                    .collect(),
            },
            r#"
books = []
[output.tex]
created = 1978
xml = false
parsing = "Imposible"
# Comment
[output.html]
output-url = "/bookshelf/"
minify = false
        "#,
        )
    }

    #[test]
    fn with_loc() {
        check_disc_conf(
            DiskConf {
                books: vec![
                    l("bare"),
                    Location::Named {
                        name: "trpl".to_owned(),
                        root_loc: "book/src/".to_owned(),
                    },
                    l("another-bare"),
                    Location::Named {
                        name: "cxx".to_owned(),
                        root_loc: "cxx/guide/src".to_owned(),
                    },
                    l("xmark"),
                    l("dummy"),
                ],
                output: Table::new(),
            },
            r#"
books = [
    "bare",
    { location = "book/src/", name = "trpl" },
    "another-bare",
    { name = "cxx", location="cxx/guide/src" },
    "xmark",
    "dummy",
]
[output]
        "#,
        )
    }
}
