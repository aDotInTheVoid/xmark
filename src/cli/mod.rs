// SPDX-License-Identifier: GPL-3.0-only
use clap::Clap;
use eyre::Result;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tracing::instrument;

pub mod config;
pub mod summary;

mod create_missing;

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Args {
    pub dir: PathBuf,
    pub create: bool,
    pub templates: Option<PathBuf>,
}

impl Args {
    #[instrument]
    pub fn parse() -> Result<Self> {
        Self::parse_from(ArgsInner::parse())
    }

    #[instrument]
    fn parse_from(inner: ArgsInner) -> Result<Self> {
        let ArgsInner {
            mut dir,
            create,
            mut templates,
        } = inner;
        dir = dir.canonicalize()?;
        templates = match templates {
            Some(p) => Some(p.canonicalize()?),
            None => None,
        };
        Ok(Args {
            dir,
            create,
            templates,
        })
    }
}

#[instrument]
pub fn init() -> Result<()> {
    color_eyre::install()?;

    use tracing_error::ErrorLayer;
    use tracing_subscriber::prelude::*;
    use tracing_subscriber::{fmt, EnvFilter};

    let fmt_layer = fmt::layer().with_target(false);
    let filter_layer = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("info"))
        .unwrap();

    tracing_subscriber::registry()
        .with(filter_layer)
        .with(fmt_layer)
        .with(ErrorLayer::default())
        .init();

    Ok(())
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: PathBuf,
    #[clap(long)]
    pub create: bool,
    #[clap(short, long)]
    pub templates: Option<PathBuf>,
}

#[cfg(test)]
mod tests {
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::test_utils::manifest_dir_redacter;

    use super::*;
    #[test]
    fn canonicalize_cwd_dir() {
        let args = ArgsInner {
            dir: ".".into(),
            create: true,
            templates: None,
        };
        let args = Args::parse_from(args).unwrap();
        // Yaml uses ~ for null
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }

    #[test]
    fn canonicalize_rel_templ() {
        let args = ArgsInner {
            dir: ".".into(),
            create: true,
            // This needs to exist, as `.canonicalize` will read, in case it's a symblink
            templates: Some("./www".into()),
        };
        let args = Args::parse_from(args).unwrap();
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }

    #[test]
    fn absolute() {
        let args = ArgsInner {
            // This should exist on most non windows systems
            // TODO: win
            dir: "/usr".into(),
            create: false,
            templates: None,
        };
        let args = Args::parse_from(args).unwrap();
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }
}
