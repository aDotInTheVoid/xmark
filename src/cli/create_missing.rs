// SPDX-License-Identifier: GPL-3.0-only
use super::summary::{Link, Summary};
use eyre::Result;
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

use log::debug;

pub fn create_missing(src_dir: &Path, summary: &Summary) -> Result<()> {
    let mut items: Vec<_> = summary
        .prefix_chapters
        .iter()
        .chain(summary.suffix_chapters.iter())
        .map(|chapter| Link {
            // This is like 3 strings
            chapter: chapter.clone(),
            ..Default::default()
        })
        //TODO: Don't clone, it can be quite nested
        .chain(summary.numbered_chapters.iter().cloned())
        .collect();

    while !items.is_empty() {
        let link = items.pop().expect("already checked");

        if let Some(ref location) = link.chapter.location {
            let filename = src_dir.join(location);
            if !filename.exists() {
                if let Some(parent) = filename.parent() {
                    if !parent.exists() {
                        fs::create_dir_all(parent)?;
                    }
                }
                debug!("Creating missing file {}", filename.display());

                let mut f = File::create(&filename)?;
                writeln!(f, "# {}", link.chapter.name)?;
            }
        }

        items.extend(link.nested_items);
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeSet;
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::cli::{self, config};
    use crate::html_render;

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();

        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "*/SUMMARY.md"],
        )
        .unwrap();

        let args = cli::Args {
            dir: temp.path().to_owned(),
            create: true,
            ..Default::default()
        };

        let conf = config::load(&args).unwrap();
        let render = html_render::HTMLRender::new(&conf, &args).unwrap();
        render.render().unwrap();

        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .filter(|x| !x.contains("_out"))
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }
}
