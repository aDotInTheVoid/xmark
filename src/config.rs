use std::fs;
use std::path::PathBuf;

use eyre::{Result, WrapErr};
use serde::{Deserialize, Serialize};

use crate::cli;
use crate::create_missing::create_missing;
use crate::summary::{parse_summary, Chapter, Summary};

/// The Config as represented in the global xmark.toml
#[derive(Clone, Debug, Hash, Serialize, Deserialize)]
pub struct GlobalConfigRepr {
    pub books: Vec<String>,
}

/// The config as usable for the programm
#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct GlobalConf {
    pub books: Vec<BookConf>,
}

#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct BookConf {
    pub location: PathBuf,
    pub slug: String,
    pub summary: Summary,
}

pub fn load(args: &cli::Args) -> Result<GlobalConf> {
    let conf = fs::read_to_string(args.dir.clone().join("xmark.toml"))
        .with_context(|| "Couldn't find xmark.toml")?;
    let conf: GlobalConfigRepr = toml::from_str(&conf)?;

    hydrate(conf, &args)
}

// Convert the disk format to a usable form
pub fn hydrate(gcr: GlobalConfigRepr, args: &cli::Args) -> Result<GlobalConf> {
    Ok(GlobalConf {
        books: gcr
            .books
            .iter()
            .map(|name| {
                let location = args.dir.join(name);
                let summary_location = location.join("SUMMARY.md");
                let mut summary = parse_summary(
                    &fs::read_to_string(&summary_location)
                        .wrap_err_with(|| format!("Couldn't open {:?}", summary_location))?,
                )?;

                if args.create {
                    create_missing(&location, &summary)?;
                }

                let fix_chap_loc = |chap: &mut Chapter| {
                    if let Some(loc) = chap.location.as_deref() {
                        chap.location = Some(location.join(loc));
                    }
                };

                summary.prefix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.suffix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.numbered_chapters.iter_mut().for_each(|chap| {
                    chap.map_mut(fix_chap_loc);
                });

                Ok(BookConf {
                    location,
                    summary,
                    slug: name.to_owned(),
                })
            })
            .collect::<Result<_>>()?,
    })
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn de_gloal_config() {
        let inp = "books = [
            'book-1',
            'book-2',
            'book-3',
            ]
";
        let conf: GlobalConfigRepr = toml::from_str(inp).unwrap();
        assert_eq!(conf.books, vec!["book-1", "book-2", "book-3"]);
    }

    #[test]
    fn hydrate_basic() {
        let args = cli::Args {
            dir: "/home/etc/bax".into(),
            create: false,
        };
        let gcr = GlobalConfigRepr { books: vec![] };
        let gc = GlobalConf { books: vec![] };
        assert_eq!(hydrate(gcr, &args).unwrap(), gc);
    }

    #[test]
    fn hydrate_dummy() {
        // I don't know why I need the anotation
        let redacter = |mut val, _: insta::internals::ContentPath| {
            // TODO: Dont do insta crimes.
            // This is in #[doc(Hidden)] internals.
            while let insta::internals::Content::Some(some) = val {
                val = *some;
            }

            if let insta::internals::Content::String(s) = val {
                s.replace(dbg!(env!("CARGO_MANIFEST_DIR")), "BASEDIR")
            } else {
                "".into()
            }
        };

        let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book");
        let args = cli::Args { dir, create: false };
        let conf = load(&args).unwrap();
        assert_yaml_snapshot!(conf, {
            ".location" => insta::dynamic_redaction(redacter),
            ".**.location" => insta::dynamic_redaction(redacter),
        });
    }
}
