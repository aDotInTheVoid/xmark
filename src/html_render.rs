use std::fs;
use std::io::Write;
use std::path::Path;

use eyre::{Context, Result};
use pulldown_cmark::{html, Options, Parser};

use crate::cli;
use crate::config::BookConf;
use crate::summary::Chapter;

pub fn render(book: BookConf, args: &cli::Args) -> Result<()> {
    let out_dir = args.dir.clone().join("_out").join("html");
    fs::create_dir_all(&out_dir)?;

    for i in book.summary.prefix_chapters {
        render_chap_io(&i, &out_dir, &args.dir)?;
    }
    for i in book.summary.suffix_chapters {
        render_chap_io(&i, &out_dir, &args.dir)?;
    }
    for i in book.summary.numbered_chapters {
        i.try_map(|chap| render_chap_io(chap, &out_dir, &args.dir))?;
    }

    Ok(())
}

fn render_chap_io(chapter: &Chapter, build_dir: &Path, base_dir: &Path) -> Result<()> {
    // Not a draft
    if let Some(ref loc) = chapter.location {
        let content = fs::read_to_string(loc)?;
        let html = render_chap(&content);

        let mut path = build_dir.join(loc.strip_prefix(base_dir).expect("Unreachble"));

        path.set_extension("html");

        fs::create_dir_all(path.parent().unwrap())?;

        let mut file =
            fs::File::create(&path).wrap_err_with(|| format!("Failed to create {:?}", &path))?;
        file.write_all(html.as_bytes())?;
    }

    Ok(())
}

// Pure inner for testing
fn render_chap(content: &str) -> String {
    let opts = Options::all();
    let parser = Parser::new_ext(content, opts);
    let mut out = String::new();
    html::push_html(&mut out, parser);
    out
}

#[cfg(test)]
mod tests {
    use std::collections::{BTreeSet};
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::{cli, config, html_render};

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "book-*/**"],
        )
        .unwrap();

        let args = cli::Args {
            create: false,
            dir: temp.path().to_owned(),
        };
        let conf = config::load(&args).unwrap();
        for book in conf.books {
            html_render::render(book, &args).unwrap()
        }


        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }
}
