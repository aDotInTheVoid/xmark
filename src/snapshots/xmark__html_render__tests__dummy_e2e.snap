---
source: src/html_render.rs
expression: paths
---
- ""
- /_out
- /_out/html
- /_out/html/book-1
- /_out/html/book-1/appendix-00.html
- /_out/html/book-1/appendix-01-keywords.html
- /_out/html/book-1/appendix-02-operators.html
- /_out/html/book-1/appendix-03-derivable-traits.html
- /_out/html/book-1/appendix-04-useful-development-tools.html
- /_out/html/book-1/appendix-05-editions.html
- /_out/html/book-1/appendix-06-translation.html
- /_out/html/book-1/appendix-07-nightly-rust.html
- /_out/html/book-1/ch00-00-introduction.html
- /_out/html/book-1/ch01-00-getting-started.html
- /_out/html/book-1/ch01-01-installation.html
- /_out/html/book-1/ch01-02-hello-world.html
- /_out/html/book-1/ch01-03-hello-cargo.html
- /_out/html/book-1/ch02-00-guessing-game-tutorial.html
- /_out/html/book-1/ch03-00-common-programming-concepts.html
- /_out/html/book-1/ch03-01-variables-and-mutability.html
- /_out/html/book-1/ch03-02-data-types.html
- /_out/html/book-1/ch03-03-how-functions-work.html
- /_out/html/book-1/ch03-04-comments.html
- /_out/html/book-1/ch03-05-control-flow.html
- /_out/html/book-1/ch04-00-understanding-ownership.html
- /_out/html/book-1/ch04-01-what-is-ownership.html
- /_out/html/book-1/ch04-02-references-and-borrowing.html
- /_out/html/book-1/ch04-03-slices.html
- /_out/html/book-1/ch05-00-structs.html
- /_out/html/book-1/ch05-01-defining-structs.html
- /_out/html/book-1/ch05-02-example-structs.html
- /_out/html/book-1/ch05-03-method-syntax.html
- /_out/html/book-1/ch06-00-enums.html
- /_out/html/book-1/ch06-01-defining-an-enum.html
- /_out/html/book-1/ch06-02-match.html
- /_out/html/book-1/ch06-03-if-let.html
- /_out/html/book-1/ch07-00-managing-growing-projects-with-packages-crates-and-modules.html
- /_out/html/book-1/ch07-01-packages-and-crates.html
- /_out/html/book-1/ch07-02-defining-modules-to-control-scope-and-privacy.html
- /_out/html/book-1/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.html
- /_out/html/book-1/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html
- /_out/html/book-1/ch07-05-separating-modules-into-different-files.html
- /_out/html/book-1/ch08-00-common-collections.html
- /_out/html/book-1/ch08-01-vectors.html
- /_out/html/book-1/ch08-02-strings.html
- /_out/html/book-1/ch08-03-hash-maps.html
- /_out/html/book-1/ch09-00-error-handling.html
- /_out/html/book-1/ch09-01-unrecoverable-errors-with-panic.html
- /_out/html/book-1/ch09-02-recoverable-errors-with-result.html
- /_out/html/book-1/ch09-03-to-panic-or-not-to-panic.html
- /_out/html/book-1/ch10-00-generics.html
- /_out/html/book-1/ch10-01-syntax.html
- /_out/html/book-1/ch10-02-traits.html
- /_out/html/book-1/ch10-03-lifetime-syntax.html
- /_out/html/book-1/ch11-00-testing.html
- /_out/html/book-1/ch11-01-writing-tests.html
- /_out/html/book-1/ch11-02-running-tests.html
- /_out/html/book-1/ch11-03-test-organization.html
- /_out/html/book-1/ch12-00-an-io-project.html
- /_out/html/book-1/ch12-01-accepting-command-line-arguments.html
- /_out/html/book-1/ch12-02-reading-a-file.html
- /_out/html/book-1/ch12-03-improving-error-handling-and-modularity.html
- /_out/html/book-1/ch12-04-testing-the-librarys-functionality.html
- /_out/html/book-1/ch12-05-working-with-environment-variables.html
- /_out/html/book-1/ch12-06-writing-to-stderr-instead-of-stdout.html
- /_out/html/book-1/ch13-00-functional-features.html
- /_out/html/book-1/ch13-01-closures.html
- /_out/html/book-1/ch13-02-iterators.html
- /_out/html/book-1/ch13-03-improving-our-io-project.html
- /_out/html/book-1/ch13-04-performance.html
- /_out/html/book-1/ch14-00-more-about-cargo.html
- /_out/html/book-1/ch14-01-release-profiles.html
- /_out/html/book-1/ch14-02-publishing-to-crates-io.html
- /_out/html/book-1/ch14-03-cargo-workspaces.html
- /_out/html/book-1/ch14-04-installing-binaries.html
- /_out/html/book-1/ch14-05-extending-cargo.html
- /_out/html/book-1/ch15-00-smart-pointers.html
- /_out/html/book-1/ch15-01-box.html
- /_out/html/book-1/ch15-02-deref.html
- /_out/html/book-1/ch15-03-drop.html
- /_out/html/book-1/ch15-04-rc.html
- /_out/html/book-1/ch15-05-interior-mutability.html
- /_out/html/book-1/ch15-06-reference-cycles.html
- /_out/html/book-1/ch16-00-concurrency.html
- /_out/html/book-1/ch16-01-threads.html
- /_out/html/book-1/ch16-02-message-passing.html
- /_out/html/book-1/ch16-03-shared-state.html
- /_out/html/book-1/ch16-04-extensible-concurrency-sync-and-send.html
- /_out/html/book-1/ch17-00-oop.html
- /_out/html/book-1/ch17-01-what-is-oo.html
- /_out/html/book-1/ch17-02-trait-objects.html
- /_out/html/book-1/ch17-03-oo-design-patterns.html
- /_out/html/book-1/ch18-00-patterns.html
- /_out/html/book-1/ch18-01-all-the-places-for-patterns.html
- /_out/html/book-1/ch18-02-refutability.html
- /_out/html/book-1/ch18-03-pattern-syntax.html
- /_out/html/book-1/ch19-00-advanced-features.html
- /_out/html/book-1/ch19-01-unsafe-rust.html
- /_out/html/book-1/ch19-03-advanced-traits.html
- /_out/html/book-1/ch19-04-advanced-types.html
- /_out/html/book-1/ch19-05-advanced-functions-and-closures.html
- /_out/html/book-1/ch19-06-macros.html
- /_out/html/book-1/ch20-00-final-project-a-web-server.html
- /_out/html/book-1/ch20-01-single-threaded.html
- /_out/html/book-1/ch20-02-multithreaded.html
- /_out/html/book-1/ch20-03-graceful-shutdown-and-cleanup.html
- /_out/html/book-1/foreword.html
- /_out/html/book-1/title-page.html
- /_out/html/book-2
- /_out/html/book-2/README.html
- /_out/html/book-2/cli
- /_out/html/book-2/cli/README.html
- /_out/html/book-2/cli/build.html
- /_out/html/book-2/cli/clean.html
- /_out/html/book-2/cli/init.html
- /_out/html/book-2/cli/serve.html
- /_out/html/book-2/cli/test.html
- /_out/html/book-2/cli/watch.html
- /_out/html/book-2/continuous-integration.html
- /_out/html/book-2/for_developers
- /_out/html/book-2/for_developers/README.html
- /_out/html/book-2/for_developers/backends.html
- /_out/html/book-2/for_developers/preprocessors.html
- /_out/html/book-2/format
- /_out/html/book-2/format/README.html
- /_out/html/book-2/format/config.html
- /_out/html/book-2/format/mathjax.html
- /_out/html/book-2/format/mdbook.html
- /_out/html/book-2/format/summary.html
- /_out/html/book-2/format/theme
- /_out/html/book-2/format/theme/README.html
- /_out/html/book-2/format/theme/editor.html
- /_out/html/book-2/format/theme/index-hbs.html
- /_out/html/book-2/format/theme/syntax-highlighting.html
- /_out/html/book-2/misc
- /_out/html/book-2/misc/contributors.html
- /_out/html/book-3
- /_out/html/book-3/c1.html
- /_out/html/book-3/c2.html
- /_out/html/book-3/c3.html
- /_out/html/book-3/c31.html
- /_out/html/book-3/c32.html
- /_out/html/book-3/c33.html
- /_out/html/book-3/post1.html
- /_out/html/book-3/pre1.html
- /_out/html/book-3/pre2.html
- /book-1
- /book-1/SUMMARY.md
- /book-1/appendix-00.md
- /book-1/appendix-01-keywords.md
- /book-1/appendix-02-operators.md
- /book-1/appendix-03-derivable-traits.md
- /book-1/appendix-04-useful-development-tools.md
- /book-1/appendix-05-editions.md
- /book-1/appendix-06-translation.md
- /book-1/appendix-07-nightly-rust.md
- /book-1/ch00-00-introduction.md
- /book-1/ch01-00-getting-started.md
- /book-1/ch01-01-installation.md
- /book-1/ch01-02-hello-world.md
- /book-1/ch01-03-hello-cargo.md
- /book-1/ch02-00-guessing-game-tutorial.md
- /book-1/ch03-00-common-programming-concepts.md
- /book-1/ch03-01-variables-and-mutability.md
- /book-1/ch03-02-data-types.md
- /book-1/ch03-03-how-functions-work.md
- /book-1/ch03-04-comments.md
- /book-1/ch03-05-control-flow.md
- /book-1/ch04-00-understanding-ownership.md
- /book-1/ch04-01-what-is-ownership.md
- /book-1/ch04-02-references-and-borrowing.md
- /book-1/ch04-03-slices.md
- /book-1/ch05-00-structs.md
- /book-1/ch05-01-defining-structs.md
- /book-1/ch05-02-example-structs.md
- /book-1/ch05-03-method-syntax.md
- /book-1/ch06-00-enums.md
- /book-1/ch06-01-defining-an-enum.md
- /book-1/ch06-02-match.md
- /book-1/ch06-03-if-let.md
- /book-1/ch07-00-managing-growing-projects-with-packages-crates-and-modules.md
- /book-1/ch07-01-packages-and-crates.md
- /book-1/ch07-02-defining-modules-to-control-scope-and-privacy.md
- /book-1/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.md
- /book-1/ch07-04-bringing-paths-into-scope-with-the-use-keyword.md
- /book-1/ch07-05-separating-modules-into-different-files.md
- /book-1/ch08-00-common-collections.md
- /book-1/ch08-01-vectors.md
- /book-1/ch08-02-strings.md
- /book-1/ch08-03-hash-maps.md
- /book-1/ch09-00-error-handling.md
- /book-1/ch09-01-unrecoverable-errors-with-panic.md
- /book-1/ch09-02-recoverable-errors-with-result.md
- /book-1/ch09-03-to-panic-or-not-to-panic.md
- /book-1/ch10-00-generics.md
- /book-1/ch10-01-syntax.md
- /book-1/ch10-02-traits.md
- /book-1/ch10-03-lifetime-syntax.md
- /book-1/ch11-00-testing.md
- /book-1/ch11-01-writing-tests.md
- /book-1/ch11-02-running-tests.md
- /book-1/ch11-03-test-organization.md
- /book-1/ch12-00-an-io-project.md
- /book-1/ch12-01-accepting-command-line-arguments.md
- /book-1/ch12-02-reading-a-file.md
- /book-1/ch12-03-improving-error-handling-and-modularity.md
- /book-1/ch12-04-testing-the-librarys-functionality.md
- /book-1/ch12-05-working-with-environment-variables.md
- /book-1/ch12-06-writing-to-stderr-instead-of-stdout.md
- /book-1/ch13-00-functional-features.md
- /book-1/ch13-01-closures.md
- /book-1/ch13-02-iterators.md
- /book-1/ch13-03-improving-our-io-project.md
- /book-1/ch13-04-performance.md
- /book-1/ch14-00-more-about-cargo.md
- /book-1/ch14-01-release-profiles.md
- /book-1/ch14-02-publishing-to-crates-io.md
- /book-1/ch14-03-cargo-workspaces.md
- /book-1/ch14-04-installing-binaries.md
- /book-1/ch14-05-extending-cargo.md
- /book-1/ch15-00-smart-pointers.md
- /book-1/ch15-01-box.md
- /book-1/ch15-02-deref.md
- /book-1/ch15-03-drop.md
- /book-1/ch15-04-rc.md
- /book-1/ch15-05-interior-mutability.md
- /book-1/ch15-06-reference-cycles.md
- /book-1/ch16-00-concurrency.md
- /book-1/ch16-01-threads.md
- /book-1/ch16-02-message-passing.md
- /book-1/ch16-03-shared-state.md
- /book-1/ch16-04-extensible-concurrency-sync-and-send.md
- /book-1/ch17-00-oop.md
- /book-1/ch17-01-what-is-oo.md
- /book-1/ch17-02-trait-objects.md
- /book-1/ch17-03-oo-design-patterns.md
- /book-1/ch18-00-patterns.md
- /book-1/ch18-01-all-the-places-for-patterns.md
- /book-1/ch18-02-refutability.md
- /book-1/ch18-03-pattern-syntax.md
- /book-1/ch19-00-advanced-features.md
- /book-1/ch19-01-unsafe-rust.md
- /book-1/ch19-03-advanced-traits.md
- /book-1/ch19-04-advanced-types.md
- /book-1/ch19-05-advanced-functions-and-closures.md
- /book-1/ch19-06-macros.md
- /book-1/ch20-00-final-project-a-web-server.md
- /book-1/ch20-01-single-threaded.md
- /book-1/ch20-02-multithreaded.md
- /book-1/ch20-03-graceful-shutdown-and-cleanup.md
- /book-1/foreword.md
- /book-1/title-page.md
- /book-2
- /book-2/README.md
- /book-2/SUMMARY.md
- /book-2/cli
- /book-2/cli/README.md
- /book-2/cli/build.md
- /book-2/cli/clean.md
- /book-2/cli/init.md
- /book-2/cli/serve.md
- /book-2/cli/test.md
- /book-2/cli/watch.md
- /book-2/continuous-integration.md
- /book-2/for_developers
- /book-2/for_developers/README.md
- /book-2/for_developers/backends.md
- /book-2/for_developers/preprocessors.md
- /book-2/format
- /book-2/format/README.md
- /book-2/format/config.md
- /book-2/format/mathjax.md
- /book-2/format/mdbook.md
- /book-2/format/summary.md
- /book-2/format/theme
- /book-2/format/theme/README.md
- /book-2/format/theme/editor.md
- /book-2/format/theme/index-hbs.md
- /book-2/format/theme/syntax-highlighting.md
- /book-2/misc
- /book-2/misc/contributors.md
- /book-3
- /book-3/SUMMARY.md
- /book-3/c1.md
- /book-3/c2.md
- /book-3/c3.md
- /book-3/c31.md
- /book-3/c32.md
- /book-3/c33.md
- /book-3/post1.md
- /book-3/pre1.md
- /book-3/pre2.md
- /xmark.toml
